File: ruby/binary/binary.rb
require 'awesome_print'

class Binary
  def initialize(num_bin)
    raise ArgumentError if (num_bin =~ /^(1|0)+$/).nil?
    @num_bin = num_bin
  end

  def to_decimal
    @num_bin.reverse.chars.map.with_index do |char, i|
      if char == "1"
        2**i
      else
        0
      end
    end.reduce(:+)
  end
end

module BookKeeping
  VERSION = 2
end

File: ecmascript/binary/binary.js
export default class Binary {
  constructor(num){
    this.num = String(num);
  }
  toDecimal(){
    if (!this.isBinary(this.num)) {return 0;}

    let num = [...this.num].reverse();
    return num.reduce((o,v,i) => {
      if (v === "1") {o += Math.pow(2, i);}
      return o;
    }, 0);
  }
  isBinary(n){
    let re = /^[01]+$/;
    return re.test(n);
  }
}
